torchtune/models/qwen2/_tokenizer.py

```diff
@@ -0,0 +1,353 @@
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import json
+import unicodedata
+from functools import lru_cache
+from typing import Dict, List, Optional, Tuple
+
+import regex as re
+
+from torchtune.data import Message, truncate
+from torchtune.modules.tokenizers import ModelTokenizer
+
+PRETOKENIZE_REGEX = (
+    r"(?i:'s|'t|'re|'ve|'m|'ll|'d)|"
+    r"[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"
+)
+
+QWEN2_SPECIAL_TOKENS = {
+    "<|endoftext|>": 151643,
+    "<|im_start|>": 151644,
+    "<|im_end|>": 151645,
+}
+
+
+ENDOFTEXT = "<|endoftext|>"
+IM_START = "<|im_start|>"
+IM_END = "<|im_end|>"
+
+
+@lru_cache()
+def bytes_to_unicode():
+    """
+    Returns list of utf-8 byte and a mapping to unicode strings. We specifically avoid mapping to whitespace/control
+    characters the bpe code barfs on.
+
+    The reversible bpe codes work on unicode strings. This means you need a large # of unicode characters in your vocab
+    if you want to avoid UNKs. When you're at something like a 10B token dataset you end up needing around 5K for
+    decent coverage. This is a significant percentage of your normal, say, 32K bpe vocab. To avoid that, we want lookup
+    tables between utf-8 bytes and unicode strings.
+    """
+    bs = (
+        list(range(ord("!"), ord("~") + 1))
+        + list(range(ord("¡"), ord("¬") + 1))
+        + list(range(ord("®"), ord("ÿ") + 1))
+    )
+    cs = bs[:]
+    n = 0
+    for b in range(2**8):
+        if b not in bs:
+            bs.append(b)
+            cs.append(2**8 + n)
+            n += 1
+    cs = [chr(n) for n in cs]
+    return dict(zip(bs, cs))
+
+
+def get_pairs(word):
+    """
+    Return set of symbol pairs in a word.
+
+    Word is represented as tuple of symbols (symbols being variable-length strings).
+    """
+    pairs = set()
+    prev_char = word[0]
+    for char in word[1:]:
+        pairs.add((prev_char, char))
+        prev_char = char
+    return pairs
+
+
+class Qwen2Tokenizer(ModelTokenizer):
+    """This class construct a "fast" Qwen2 tokenizer (backed by HuggingFace's *tokenizers* library).
+
+    See <https://github.com/huggingface/transformers/blob/v4.40.1/src/transformers/models/qwen2/tokenization_qwen2_fast.py>.
+
+    Args:
+        vocab_file (str): Path to vocab.json file.
+        merges_file (str): Path to merges.txt file.
+
+    Example:
+        >>> tokenizer = Qwen2Tokenizer(vocab_file="/path/to/vocab.json", merges_file="/path/to/merges.txt")
+        >>> tokenized_text = tokenizer.encode("Hello world!")
+        >>> print(tokenized_text)
+        []
+    """
+
+    system = f"{IM_START}system\n{{content}}{IM_END}\n"
+    user = f"{IM_START}user\n{{content}}{IM_END}\n"
+    assistant = f"{IM_START}assistant\n{{content}}{IM_END}\n"
+    assistant_for_generation = f"{IM_START}assistant\n"
+
+    def __init__(
```

Comment:
Is it worth thinking about trying to unify the constructor args here to help all our tokenizers look and feel the same?

Reply:
Can I set default value of `merges_file` to `None`, and infer it from `vocab_file` (`merges_file = merges_file or os.path.join(os.path.dirname(vocab_file), 'merges.txt')`)?

Some users may change `eos_token` and other special tokens when tuning Qwen2, so I keep these keyword arguments for flexibility.

Reply:
HF will always pull the `merges_file` into the same directory as the vocabulary right? That seems like a sensible suggestion, particularly if you think there are cases when a tokenizer comes with a `merges_file`, but a user wouldn't want to use it. If that's never going to be the case we could maybe just remove it from the `__init__` and hardcode loading it if it exists.

Could we also change `vocab_file` to just `path`, which means we can specify the config for this tokenizer the same as we would any other?

This is a slight tangent, and correct me if I'm wrong here, but the `tokenizer.json` format is for the fast fast BPE format, and from a quick look around, we don't support this yet for Llama models, right? Is it worth opening a broader discussion about generalising this tokenizer?

I think we also have some precedent for generally specifying custom tokens (like `eos_id`) in the codebase. cc @joecummings for thoughts here (and to generally make sure I'm not tripping).